packages/vite/src/node/server/transformRequest.ts

```diff
@@ -231,42 +289,100 @@ async function loadAndTransform(
   const mod = await moduleGraph.ensureEntryFromUrl(url, ssr)
   ensureWatchedFile(watcher, mod.file, root)
 
-  // transform
-  const transformStart = isDebug ? performance.now() : 0
-  const transformResult = await pluginContainer.transform(code, id, {
-    inMap: map,
-    ssr
-  })
-  const originalCode = code
-  if (
-    transformResult == null ||
-    (isObject(transformResult) && transformResult.code == null)
-  ) {
-    // no transform applied, keep code as-is
-    isDebug &&
-      debugTransform(
-        timeFrom(transformStart) + colors.dim(` [skipped] ${prettyUrl}`)
-      )
-  } else {
-    isDebug && debugTransform(`${timeFrom(transformStart)} ${prettyUrl}`)
-    code = transformResult.code!
-    map = transformResult.map
-  }
+  let result: TransformResult | null = null
 
-  if (map && mod.file) {
-    map = (typeof map === 'string' ? JSON.parse(map) : map) as SourceMap
-    if (map.mappings && !map.sourcesContent) {
-      await injectSourcesContent(map, mod.file, logger)
+  // persistent cache
+
+  const persistentCacheKey = includedInPersistentCache
+    ? (_persistentCache.getKey(id + code) ?? '') + (options.ssr ? '-ssr' : '')
+    : ''
+
+  const finalIncludedInPersistentCache =
+    includedInPersistentCache &&
+    !code.includes('import.meta.glob') &&
+    // TODO: handle ENV changes
+    !code.includes('import.meta.env')
```

Comment:
I need to check but after a second though I think import.meta.env is safe because it's not transformed but evaluated at runtime. The replacement of `esbuild.define` on the other hand is an issue. But maybe we could just log a warning if it's detected in the config saying that this can lead to cache issues and handle it later on.

Reply:
Maybe we can put the value of `define` into the cache version hash?